GooglePlayPlugins/com.google.play.billing/Samples/TrivialKart/scripts/Controller/PurchaseController.cs

```diff
@@ -331,6 +331,14 @@ public void OnPurchaseFailed(Product product, PurchaseFailureReason failureReaso
         // this reason with the user to guide their troubleshooting actions.
         Debug.Log(
             $"OnPurchaseFailed: FAIL. Product: '{product.definition.storeSpecificId}', PurchaseFailureReason: {failureReason}");
+
+        // When purchase fail due to duplicate transaction, and the item is not shown in client side,
+        // do fetch additional products to fetch the product.
+        // This situation may happen when the user lose internet connection after paying to play store. 
+        if (failureReason == PurchaseFailureReason.DuplicateTransaction)
```

Comment:
is this what this api is meant for? this api fetches product metadata (more like GetSkuDetails) - a more suitable use case seems to be: once user reaches a level / purchases a subs and unlocks more SKUs, they can use this api to fetch product metadata for the newly unlocked items.

The scenario you described sounds like a use case for RestoreTransactions. @Qing451800 what do you think?

Reply:
This will trigger RetrieveProducts internally and then it will trigger getPurchases.

We have done a fix such that, if we find the user is purchasing something he already owns, then we will push a library update to the clients. Triggering getPurchases will notify Unity IAP to receive this item.

You're right. RestoreTranscation does the same thing by calling getPurchases. I think we shall use RestoreTranscation here, which saves a useless skuDetails RPC call.

Reply:
I saw on the design doc you wrote that the RestoreTransaction is user-trigger, will the change break it?

Reply:
Good point. That leaves user an option to trigger this API call.

At this place, calling RestoreTransaction or FetchAdditionalProducts will both get user outside of this state. Considering that RestoreTransaction is cheaper and fits better into this context. I'm leaning to use RestoreTransaction.

Reply:
I also noticed one thing is that RestoreTransaction will go through all the currently active transactions the user has made. However, FetchAdditionalProducts will only fetch a specific product in this way. Will it also be considered?

Reply:
Does it trigger ProcessPurchases for all active transactions? If not, then it is expected.

Reply:
It will not trigger ProcessPurchases for all active transactions. For those transactions that already recorded at client side, Unity will complain "Already recorded transaction" and not do ProcessPurchase for those transactions.